### PR TITLE
Allow returning local estimators from `expect` and `expect_and_grad`

### PR DESCRIPTION
**Pull request by femtobit**
This PR introduces an (experimental) option `_return_estimators` for `MCState.expect` and `MCState.expect_and_grad`, which optionally returns the local estimators `O_loc(s) = <s|O|ψ>/<s|ψ>` and not just their statistics.

This is necessary for implementing #1140 and potentially other VMC regularization methods or preconditioners and also can be useful for debugging purposes (e.g., to check MCMC chain properties beyond those contained in `statistics`).

### CODE REVIEW

**Comment by attila-i-szabo** [comment]
Shouldn't we just add the local estimators to the `Stats` object? It's not a lot of memory and would keep things tidier

**Comment by PhilipVinc** [comment]
While I see the logic, it would defeat the purpose of a `Stats` object...

**Comment by femtobit** [comment]
I think I like the idea of including the estimators as an optional field in the stats. I wonder if stats should just contain a dictionary (like `additional_data`) where the estimators and potentially other optional quantities we may want to add at some point in the future can be stored.

**Comment by PhilipVinc** [comment]
I think hiding this stuff (and eventually other data) optionally inside of a stats object would be very wrong.
Then, when you have a stats object, you don't know if it's what you want or not.
You need to check if inside it has the energies.

It's bad UI for users as well. How do you tell them that a function failed because they provided the right object but that object was not computed with the right flag?

Isn't it so much better to have explicit energies, and then users have to explicitly manage them?

**Comment by femtobit** [comment]
> It's bad UI for users as well. How do you tell them that a function failed because they provided the right object but that object was not computed with the right flag?

That I strongly agree with. I think the stats object gets passed around to much (like to the logger, where the estimators have no place [unless users put them there explicitly to log them]).

**Comment by attila-i-szabo** [comment]
> I think now that the stats object gets passed around to much (like to the logger, where the estimators have no place [unless users put them there explicitly to log them]) for it to contain the estimators.

That's not an argument, it's not like you have to copy a large array every time you use the object somewhere

**Comment by femtobit** [comment]
> That's not an argument, it's not like you have to copy a large array every time you use the object somewhere

It is not copied, but the array does live longer if it is passed around throughout the code (instead of the memory just being handled inside the JITed function and probably reused quickly). It may only be a few megabytes at most, but still.

**Comment by femtobit** [comment]
I think for now (and for the purposes of this PR), just returning the local estimators on demand is sufficient. That way, they can be used in the driver and therefore in a potential `TDVPEquation` or more general preconditioner class.

I'd prefer keeping this part of the interface marked as experimental for now (that's why `_return_estimators` has an underscore, so people using it know that it might be subject to change in future releases). That way, we do keep options open in case we settle on a different design at some point.

**Comment by PhilipVinc** [comment]
I promise I'll look carefully and efficiently into this next week

File: netket/vqs/mc/mc_state/expect_grad.py [hunk 1]
```diff
@@ -177,8 +195,8 @@ def grad_expect_operator_kernel(
     model_state: PyTree,
     σ: jnp.ndarray,
     local_value_args: PyTree,
+    **kwargs,
```
Comment (attila-i-szabo) [inline netket/vqs/mc/mc_state/expect_grad.py:198]:
```suggestion
    _return_estimators = False
```
Is there a particular reason for doing it like this? It makes checking quite complicated

File: netket/jax/_expect.py [hunk 2]
```diff
@@ -33,33 +33,51 @@ def expect(
...
     L_σ = expected_fun(pars, σ, *expected_fun_args)
     if n_chains is not None:
         L_σ = L_σ.reshape((n_chains, -1))
 
     L̄_σ = mpi_statistics(L_σ.T)
     # L̄_σ = L_σ.mean(axis=0)
 
-    return L̄_σ.mean, L̄_σ
+    if _return_estimators:
+        return L̄_σ.mean, (L̄_σ, L_σ.T)
```
Comment (attila-i-szabo) [inline netket/jax/_expect.py:67]:
```suggestion
        return L̄_σ.mean, L̄_σ, L_σ.T
```
This output always goes through a tuple unpacking anyway, why make that more complicated?

File: netket/vqs/mc/mc_state/expect_chunked.py [hunk 3]
```diff
@@ -132,13 +145,17 @@ def logpsi(w, σ):
...
         parameters,
         σ,
         args,
         n_chains=σ_shape[0],
+        _return_estimators=_return_estimators,
     )
 
-    return Ō_stats
+    if _return_estimators:
+        return res[1], res[2]
```
Comment (attila-i-szabo) [inline netket/vqs/mc/mc_state/expect_chunked.py:159]: in fact, this only works with changing to a 3-tuple rather than `a, (b,c)`

File: netket/vqs/mc/mc_state/expect_grad.py [hunk 4]
```diff
@@ -46,25 +46,31 @@ def expect_and_grad(  # noqa: F811
...
     *,
     mutable: Any,
+    **kwargs,
 ) -> Tuple[Stats, PyTree]:
     σ, args = get_local_kernel_arguments(vstate, Ô)
 
     local_estimator_fun = get_local_kernel(vstate, Ô)
 
-    Ō, Ō_grad, new_model_state = grad_expect_hermitian(
+    res = grad_expect_hermitian(
```
Comment (attila-i-szabo) [inline netket/vqs/mc/mc_state/expect_grad.py:55]: For internal functions like `grad_expect_hermitian`, wouldn't it be easier to always return the estimators? If the wrapper function doesn't want to return them, their lifetime only extends by a few lines (if the difference isn't jitted away), but it would keep the unpacking simple

**Review by attila-i-szabo — COMMENTED** [review]
By no means a full review, just a few design things that sprung into my eye

> I think for now (and for the purposes of this PR), just returning the local estimators on demand is sufficient. That way, they can be used in the driver and therefore in a potential TDVPEquation or more general preconditioner class.

Fully agreed. In fact, `expect` and `expect_and_grad` can't do any more than this, the discussion we had was already more about how to use this interface in drivers

> I'd prefer keeping this part of the interface marked as experimental for now (that's why _return_estimators has an underscore, so people using it know that it might be subject to change in future releases). That way, we do keep options open in case we settle on a different design at some point.

Given this, I don't think we need to keep this experimental, once we figure out how this part of the logic should work, we could call the argument just `return_estimators`. Also, it would be a pain to change the argname if stuff already depends on it

**Comment by danielalcalde** [comment]
I'm a bit late to the discussion, but I would suggest that expect_and_grad would return an object similar to the quantum geometric tensor that inside has the O_loc(s) = <s|O|ψ>/<s|ψ>.

E, F = variational_state.expect_and_grad(hamiltonian)
type(F) == operator_gradient
type(F._dict) == FrozenDict # What is now returned by expect_and_grad
type(F.to_dense()) == jnp.ndarray # an unraveling of F._dict

This would make things cleaner in my opinion.

**Comment by PhilipVinc** [comment]
>I'm a bit late to the discussion, but I would suggest that expect_and_grad would return an object similar to the quantum geometric tensor that inside has the O_loc(s) = <s|O|ψ>/<s|ψ>.

This is tricky because the way to define the raveling is quite arbitrary and depends on the model so it requires additional input that you don't get at this point (while QGT gets it...)

Also, we can't change the pytree structure of `F` without breaking `tree_map` updates of the parameters.

We can hide `O_loc` inside of the expectation value/stats object but I am mostly against it because it wouldn't be an expectation value object anymore.
Also, we store and pass those objects around and if they become 'heavyweights' it might become an issue.
Though I'd love to be convinced of the contrary.

>I also would add an extra function called expect_and_grad_and_quantum_geometric_tensor that would compute both the gradient and the quantum geometric tensor at a time. This can be done more efficiently than at the moment is implemented. Right now the gradient is computed twice, one time for the geometric tensor and another time for the gradient. Merging the two led to 20% speedup in my current project.

Yes. I agree. Though I would like to see a design proposal for this... There's some edge cases

**Comment by PhilipVinc** [comment]
>You have to keep in mind that we have several different implementations of the QGT and returning the full Jacobian would be a disaster if it doesn't fit into memory but you can get on fine with QGTOnTheFly. Also, QGTJacobian doesn't just use a plain VJP, so I'm not 100% sure what could be recycled

Most likely the forward pass gets recycled. 
The backward pass cannot be reused.

**Comment by danielalcalde** [comment]
> Not sure I understand this. How is the data content of `F` different from `F._dict`? 
F would also have hidden inside of it the O_loc(s) vector. Since I agree with Filippo that saving it into the stats object is not ideal.

> This is tricky because the way to define the raveling is quite arbitrary and depends on the model so it requires additional input that you don't get at this point (while QGT gets it...)

The raveling would in this case be saved in the F object as well. One could for example have the function: F.unravel_like(dw_dt)

> I agree that recomputing gradients (or rather the VJP function) is inefficient, but what exactly would `expect_and_grad_and_quantum_geometric_tensor` return? You have to keep in mind that we have several different implementations of the QGT and returning the full Jacobian would be a disaster if it doesn't fit into memory but you can get on fine with `QGTOnTheFly`. Also, `QGTJacobian` doesn't just use a plain VJP, so I'm not 100% sure what could be recycled

You are right! The function `expect_and_grad_and_quantum_geometric_tensor` would only be useful when one uses `QGTJacobian`. But in that case it is quite a bit faster.

**Comment by femtobit** [comment]
> I'm a bit late to the discussion, but I would suggest that expect_and_grad would return an object similar to the quantum geometric tensor that inside has the O_loc(s) = <s|O|ψ>/<s|ψ>.

I agree that this could be cleaner and it would mirror our QGT interface which returns an object that, for QGTJacobian at least, contains the log-derivatives as an additional field. However, this would be a breaking change to `expect_and_grad` (which the present change is not). While we could make the return type of the gradients depend on an option too, I am not convinced the increase in elegance would be worth that effort.

> You are right! The function `expect_and_grad_and_quantum_geometric_tensor` would only be useful when one uses QGTJacobian. But in that case it is quite a bit faster.

I see the benefit of an `expect_grad_qgt` function for performance reasons in the `QGTJacobian*` cases (or at least for `QGTJacobianDense`?). In particular, the noise-based regularization #1140 requires the full spectrum and eigenvectors of the QGT, so in those cases `qgt.to_dense()` will inevitably be called. We can always have `expect_grad_qgt` fall back to call `expect_and_grad` and `quantum_geometric_tensor` internally if no more efficient code path is available.

The question is: Can we build, say, the current `QGTJacobianDense` with information from the Hermitian `expect_and_grad` hierarchy of methods without modifying the QGT class? Or would we need to special-case everything to make this work? @PhilipVinc

**Comment by femtobit** [comment]
Also, does anyone have an idea why the Windows CI tests keep hanging here? And there is the crash of `test/variational/test_variational_mixed.py::test_expect_exact[model:(R->C)-superop:(Lind^2)]`, but at least that seems fully reproducible, while the Windows CI doesn't hang in #1164, for example.

File: netket/jax/_expect.py (hunk 2 above)
Comment (femtobit) [reply netket/jax/_expect.py]: I think there was some issue with `vjp` when calling it without grouping the non-differentiated return values in a single tuple.

**Comment by danielalcalde** [comment]
> However, this would be a breaking change to expect_and_grad (which the present change is not).

Yes, I'm convinced, it would break back compatibility.

> Can we build, say, the current QGTJacobianDense with information from the Hermitian expect_and_grad hierarchy of methods without modifying the QGT class?

The way I have it implement the expect_and_grad_qgt right now is that I compute only O_loc(s) and then steal the O_k(s) from the qgt tensor to compute the gradients.

**Comment by femtobit** [comment]
I wonder if it would make sense to just have a separate method to directly compute and return the log-derivatives (i.e., the "`O_k`s"). (The local estimators can still be obtained from `expect` in any case, the additional computation of the statistics should not be a huge overhead.) Then these can be used directly in cases where manual computation of the (dense) QGT and gradients is useful and done by the "traditional" VMC formulas.

(This is somewhat orthogonal to this PR, as returning O_loc from `expect` is sensible in any case and useful not just for VMC/TDVP algorithms.)

**Comment by attila-i-szabo** [comment]
> I wonder if it would make sense to just have a separate method to directly compute and return the log-derivatives (i.e., the "`O_k`s"). (The local estimators can still be obtained from `expect` in any case, the additional computation of the statistics should not be a huge overhead.) Then these can be used directly in cases where manual computation of the (dense) QGT and gradients is useful and done by the "traditional" VMC formulas.

It makes sense to have a method like this, if only to make `MCState` fully featured, you might be interested in wave function gradients without wanting to build a QGT.

> The question is: Can we build, say, the current `QGTJacobianDense` with information from the Hermitian `expect_and_grad` hierarchy of methods without modifying the QGT class? Or would we need to special-case everything to make this work?

First off, `expect_and_grad` never computes the full Jacobian, but this isn't a big issue. The data content of `QGTJacobian*` is just the Jacobian matrix/pytree, so you could imagine an implementation of `expect_and_grad` that uses the same data to get the energy gradients.

The problem is more that everything atm is geared towards constructing QGTs from vstates. The "constructor functions" we expose to the users explicitly take a vstate, the preconditioner interface assumes one of the inputs is a vstate, etc. We could expose `QGTJacobianDenseT` directly, which would take the Jacobian matrix, but you do need to special-case everything for that to work. If we want to have this for VMC (and I want to), we'd need to special-case the whole preconditioner/SR interface and have the driver manage the Jacobian until it can be fed into the preconditioner. This sounds like a mess.

I think a better solution would be adding a `jacobian: Optional[PyTree]` field to `MCState` (I would keep it as a pytree, that's how it's first generated anyway and `Dense` is already set up to convert it to a matrix). Then, `QGTJacobian*` can check if there is something in this field, use it if yes or compute it from scratch (and memoise it in the vstate) if no. The Jacobian only depends on things that are part of the vstate, so it makes sense to keep it there. Then, the only thing that needs to be special-cased is letting the driver know it should compute the Jacobian together with `expect_and_grad`.

What do you think?

> (This is somewhat orthogonal to this PR, as returning O_loc from `expect` is sensible in any case and useful not just for VMC/TDVP algorithms.)

Maybe we should have an issue dedicated to discussing preconditioner interfaces. It's indeed not a part of this PR, it's also holding up #1142, and having a shared interface for VMC preconditioners and TDVP equations would be very important IMO.

**Comment by femtobit** [comment]
> It makes sense to have a method like this, if only to make `MCState` fully featured, you might be interested in wave function gradients without wanting to build a QGT.

Indeed. We do advertise the ability to write customs algorithms without using an existing driver class at various points in the documentation. And while this is largely true (and, for example, even without this PR you can get the local values from an `MCState` in a [couple lines of code](https://gist.github.com/femtobit/945473a25ca46abb02a9762e69302ef3)), doing so often requires non-trivial knowledge of `MCState` internals and those take a while to understand due to the many different variations w.r.t. to options like chunking, pure or mixed states, Hermitian and non-Hermitian operators, etc. we have.

As far as I'm concerned, I would propose we do the following:
1) Finish this PR by just adding `return_estimators`. This removes the need for writing more involved custom code like I have shown above and already helps, e.g., with debugging MCMC chain convergence.
2) Add a way to get the log-derivatives from `MCState`. Ideally, this should work together with `expect_and_grad` like by setting a `.jacobian` field (or maybe even just caching information that can be potentially reused by a `.jacobian()` function which can also be called independently?) as @attila-i-szabo suggested.

With these two features, implementing #1140 should already be possible. Prototyping this for `TDVP` (where due to it being `experimental` we are not yet bound by older design decisions and can start by adding a working implementation and then iterate from there) should also be helpful to understand what a good generic TDVP solver interface should look like. (I am avoiding the term preconditioner as I think swapping out the equation of motion goes conceptually beyond preconditioning.)

**Comment by attila-i-szabo** [comment]
> As far as I'm concerned, I would propose we do the following:
> 
>  1. Finish this PR by just adding `return_estimators`. This removes the need for writing more involved custom code like I have shown above and already helps, e.g., with debugging MCMC chain convergence.

Agreed, this PR is an improvement in itself and doesn't tie our hands re how we can use it.

>  2. Add a way to get the log-derivatives from `MCState`. This could work together with `expect_and_grad` by setting a `.jacobian` field (or maybe even just caching information that can be potentially reused by a `.jacobian()` function which can also be called independently?) as @attila-i-szabo suggested.

We can figure this out later, but something like the `@cached_property` pattern of netket dataclasses would be nice for this (with a way for `expect_and_grad_and_jacobian` to fill in the cache)

> With these two features, implementing #1140 should already be possible. Prototyping this for `TDVP` (where due to it being `experimental` we are not yet bound by older design decisions and can start by adding a working implementation and then iterate from there) should also be helpful to understand what a good generic TDVP solver interface should look like. 

I am half happy with this.
* It's indeed easier to come up with a working implementation without worrying about breaking changes
* OTOH, there's a risk we end up with two mutually incompatible implementations of the same algorithm, which would be annoying for future development. In this case, #1140 makes perfect sense for VMC as well, but we might end up with an implementation that would require us to tear up the VMC driver if we want to fit it in. (I'm cool with having to use a different SR class, but it should talk to the VMC driver in the same way.) This is the main reason I want to use the same preconditioner interface for VMC and TDVP, that we don't have to implement every improvement to this pattern twice.

> (I am avoiding the term preconditioner as I think swapping out the equation of motion goes conceptually beyond preconditioning.)

I understand this, but I'm generally sceptical about expressing purely conceptual differences with code. Mathematically, we still do the same thing and computational improvements for one are typically improvements for the other, so having two incompatible/independent interfaces for it would duplicate a lot of code, and neither would be fully featured in the long run. If you think it's more intuitive for TDVP users not to bundle the QGT and the solver, we can still combine them internally (OTOH, stuff like #1142 is best expressed on the level of the bundle, so you could still lose useful functionality)

**Comment by gcarleo** [comment]
My two cents on the design here: in older versions 'machines' (a.k.a.  models) had a method to compute the derivatives of the log, O_k. For reasons I don't remember this method has been removed, and only log_value is left
In my opinion, adding a convenience function to compute the O_k (independently of the QGT) would be nice though, since it would allow to do essentially everything without resorting to drivers. I would just put that convenience method in the general interface of the variational state, as it used to be. QGT then could decide to reuse this information for efficiency, if already precomputed (I guess that's what Attila is suggesting).

**Comment by danielalcalde** [comment]
> I think a better solution would be adding a `jacobian: Optional[PyTree]` field to `MCState` (I would keep it as a pytree, that's how it's first generated anyway and `Dense` is already set up to convert it to a matrix). Then, `QGTJacobian*` can check if there is something in this field, use it if yes or compute it from scratch (and memoise it in the vstate) if no. The Jacobian only depends on things that are part of the vstate, so it makes sense to keep it there. Then, the only thing that needs to be special-cased is letting the driver know it should compute the Jacobian together with `expect_and_grad`.
> 
> What do you think?

Seems like an elegant solution :)

**Comment by femtobit** [comment]
I think we are converging on a flexible design in the preconditioner PR discussion #1142.

Is there anything specific to this feature of returning the local estimators that still needs to be resolved here? Otherwise, I'd like to merge this and continue the discussion over there. Any opinions? (Or an approving review?) @PhilipVinc @gcarleo @attila-i-szabo

**Comment by femtobit** [comment]
Closed in favor of #1179.
